bioconda_utils/update_pinnings.py

```diff
@@ -1,88 +1,125 @@
-#!/usr/bin/env python
-from conda_build import api
 import re
 import sys
 import os.path
+import logging
+import collections
+import enum
+
+import networkx as nx
+
+from .utils import RepoData, load_conda_build_config, parallel_iter
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
 buildNumberRegex = re.compile("^( +number: +)(\d+)(.*?)$")
 buildNumberJinja = re.compile("^({% +set build +=.+?)(\d+)(.*?%})")
 
-def already_bumped(r, name, version, buildNumber):
-    rsub = r.loc[(r.name == name) &
-             (r.platform != "osx") &
-             (r.version == version)]
-    if rsub.shape[0] == 0:
-        # The version was never built, it doesn't explicitly require bumping
-        return True
-    if buildNumber > max(rsub.build_number):
-        # Already bumped
-        return True
-    return False
 
+def will_build_variant(meta):
+    """Check if the recipe variant will be built as currently rendered"""
+    build_numbers = RepoData().get_package_data(
+        'build_number',
+        name=meta.name(), version=meta.version(),
+        platform=['linux', 'noarch'],
+    )
+    res = all(num < meta.build_number() for num in build_numbers)
+    if res:
+        logger.debug("Package %s=%s will be built already because %s < %s)",
+                     meta.name(), meta.version(),
+                     max(build_numbers) if build_numbers else "N/A",
+                     meta.build_number())
+    return res
 
-def already_built(r, name, version, buildstring):
-    if r.loc[(r.name == name) &
-             (r.platform != "osx") &
-             (r.version == version) &
-             (r.build == buildstring)].shape[0] > 0:
-        return True
-    return False
 
+def have_variant(meta):
+    """Checks if we have an exact match to name/version/buildstring
+    """
+    res = RepoData().get_package_data(
+        name=meta.name(), version=meta.version(), build=meta.build_id(),
+        platform=['linux', 'noarch']
+    )
```

Comment:
without any `key` set, the result is `exists`. The key could also be a list (`['version', 'build'], name='htslib'`), in which case the pandas tuple iterator is returned.